File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Subset Sum/Subset Sum.cpp

class Solution {
private:
	void myHelper(vector<int> &res, vector<int> &arr, int n, int currSum, int index) {
        if (index == n) {
            res.push_back(currSum); return;
            
        }
        // Case 1: Exclude the current character and move to the next
        myHelper(res, arr, n, currSum, index + 1);

        // Case 2: Include the current character and move to the next
        myHelper(res, arr, n, currSum + arr[index], index + 1);
    }
  public:
    vector<int> subsetSums(vector<int> arr, int n) {
        // Write Your Code here
        vector<int> res;
        myHelper(res, arr, n, 0, 0);
        return res;
    }
};

File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Subset Sum/Subset Sum.java


//User function Template for Java//User function Template for Java
class Solution{
    ArrayList<Integer> subsetSums(ArrayList<Integer> arr, int N){
        // code here
        ArrayList<Integer> total = new ArrayList<>();
        for (int i = 0; i < (1 << N); i++) {
            int subsetSum = 0;
            for (int j = 0; j < N; j++) {
                if ((i & (1 << j)) > 0) {
                    subsetSum += arr.get(j);
                }
            }
            total.add(subsetSum);
        }
        return total;
    }
}